File: kt/src/commonMain/kotlin/com/github/ephemient/aoc2021/Day10.kt
package com.github.ephemient.aoc2021

class Day10(private val lines: List<String>) {
    fun part1(): Int = lines.sumBy { line ->
        val expected = mutableListOf<Char>()
        for (char in line) {
            pairs[char]?.also { expected += it }
                ?: expected.removeLastOrNull()?.takeIf { it == char }
                ?: return@sumBy part1Points.getValue(char)
        }
        0
    }

    fun part2(): Long = lines.mapNotNullTo(mutableListOf()) { line ->
        val expected = mutableListOf<Char>()
        for (char in line) {
            pairs[char]?.also { expected += it }
                ?: expected.removeLastOrNull()?.takeIf { it == char }
                ?: return@mapNotNullTo null
        }
        expected.asReversed().fold(0L) { acc, char -> 5 * acc + part2Points.getValue(char) }
    }.apply { sort() }.let { it[it.size / 2] }

    companion object {
        private val pairs = "([{<".zip(")]}>").toMap()
        private val part1Points = mapOf(')' to 3, ']' to 57, '}' to 1197, '>' to 25137)
        private val part2Points = mapOf(')' to 1, ']' to 2, '}' to 3, '>' to 4)
    }
}

File: py/aoc2021/day10.py
pairs = {"(": ")", "[": "]", "{": "}", "<": ">"}
part1_points = {")": 3, "]": 57, "}": 1197, ">": 25137}
part2_points = {")": 1, "]": 2, "}": 3, ">": 4}


def part1(lines):
    """
    >>> part1(["[({(<(())[]>[[{[]{<()<>>", "[(()[<>])]({[<{<<[]>>(", "{([(<{}[<>[]}>{[]{[(<()>", "(((({<>}<{<{<>}{[]{[]{}", "[[<[([]))<([[{}[[()]]]", "[{[{({}]{}}([{[{{{}}([]", "{<[[]]>}<{[{[{[]{()[[[]", "[<(<(<(<{}))><([]([]()", "<{([([[(<>()){}]>(<<{{", "<{([{{}}[<[[[<>{}]]]>[]]"])
    26397
    """
    total = 0
    for line in lines:
        expected = []
        for char in line.strip():
            if char in pairs:
                expected.append(pairs[char])
            elif not expected or char != expected.pop():
                total += part1_points[char]
                break
    return total


def part2(lines):
    """
    >>> part2(["[({(<(())[]>[[{[]{<()<>>", "[(()[<>])]({[<{<<[]>>(", "{([(<{}[<>[]}>{[]{[(<()>", "(((({<>}<{<{<>}{[]{[]{}", "[[<[([]))<([[{}[[()]]]", "[{[{({}]{}}([{[{{{}}([]", "{<[[]]>}<{[{[{[]{()[[[]", "[<(<(<(<{}))><([]([]()", "<{([([[(<>()){}]>(<<{{", "<{([{{}}[<[[[<>{}]]]>[]]"])
    288957
    """
    scores = []
    for line in lines:
        expected = []
        for char in line.strip():
            if char in pairs:
                expected.append(pairs[char])
            elif not expected or char != expected.pop():
                break
        else:
            score = 0
            for char in expected[::-1]:
                score = 5 * score + part2_points[char]
            scores.append(score)
    scores.sort()
    return scores[len(scores) // 2]


parts = (part1, part2)
